### ISSUE TITLE
Stack traces inside Mono do not show file/line numbers.

### ISSUE THREAD

**Issue by PJB3005** [godotengine/godot#14822]
**Godot version:**
652c98a7beb8fc8a8aad5075a673caf21994a2ba

**OS/device including version:**
Windows 10 Pro 64-bit, Mono 5.4.1

**Issue description:**
Mono doesn't seem to be ran in debug mode, so if you get an exception, the file will just be something like this: `<abcbe92421df45dd9c0b29a6fa712911>:0`. You have to pass `--debug` to the mono command line to run it in debug mode, but you can't really do that when it's embedded inside Godot.

**Steps to reproduce:**
1. Make a new C# script.
2. Make it throw an exception.
3. Try printing the exception stack trace.

```cs
public class Thrower : Node
{
    public override void _Ready()
    {
        try
        {
            throw new Exception("whoopsie!");
        }
        catch (Exception e)
        {
            GD.Print($"Uh oh: {e}");
        }
    }
}
```

**Comment by akien-mga** [comment]
cc @neikeq

**Comment by mhilbrunner** [comment]
Duplicate of https://github.com/godotengine/godot/issues/14589 (although that is a little hard to find)

**Comment by akien-mga** [comment]
Closing as duplicate of #14589 then, but I'll edit that one to add more details from this one.

**Comment by neikeq** [comment]
Not a duplicate of #14589. I think `--debug` shouldn't be needed, but I'll check when I'm back.

**Comment by PJB3005** [comment]
--debug is only a thing on the Mono executable itself. I tried looking into
this myself, and found that all --debug did on the executable was set
mono_debug_init(), but Godot already does that so I'm not sure why it's not
working.

If you look at the source code for it, however, you'll find that it says in
the a comment the initialization of debug mode should happen between
loading of corlib and your assemblies... which the regular executable also
doesn't seem to do. So I've got no clue.

One possibility is that Mono's not loading the PDB files, which I wasn't
able to verify it's actually doing.

2017-12-19 21:39 GMT+01:00 Ignacio Etcheverry <notifications@github.com>:

**Comment by PJB3005** [comment]
Commenting what I accidentally said in the other issue:

This seems to be broken on Windows, but I've had no problem with it on MacOS. Might be related.

**Comment by neikeq** [comment]
Just tested and couldn't reproduce this on Linux, so this seems to be Windows only. I can't test on Windows right now though.

**Comment by neikeq** [comment]
Could it be because we are using System MSBuild instead of Mono's? We are using it because I can't get Mono's MSBuild to work on Windows (as well as other people), so maybe the PDB is incompatible.

**Comment by akien-mga** [comment]
We have now entered release freeze for Godot 3.0 and want to focus only on release critical issues for that milestone. Therefore, we're moving this issue to the 3.1 milestone, though a fix may be made available for a 3.0.x maintenance release after it has been tested in the master branch during 3.1 development. If you consider that this issue is critical enough to warrant blocking the 3.0 release until fixed, please comment so that we can assess it more in-depth.

**Comment by neikeq** [comment]
Just checked today and the log had several "Symbol file LoadedFromMemory is not a mono symbol file" messages. Changing DebugType to portable fixed it.

**Comment by Duehok** [comment]
> Changing DebugType to portable fixed it.
What does that means? A fix will be available in next 3.0x? In 3.x? Never? Or we can already fix it easily?
Because if this issue is the one killing debugging then it really needs to get fixed ASAP. We are not in the 60s anymore, "print to console" is not enough.

**Comment by PJB3005** [comment]
I believe he may be talking about the `DebugType` property in the csproj files. I can't test right now though. If you want to try fixing it open your `.csproj` file and change the `DebugType` tags to `portable`.

Also I think we're all well aware how painful lack of debuggers is, there's a reason C# isn't marked stable yet.

**Comment by Duehok** [comment]
Changing debugtype to portable in the .csproj did not solve this issue.
Godot 3.0.2 stable, win 8.1m mono 5.4.1.7.

**Comment by neikeq** [comment]
@Duehok You probably didn't change the one with the Tools Configuration condition. Make sure you change all them.

**Comment by PJB3005** [comment]
I can confirm personally that `<DebugType>portable</DebugType>` makes file & line numbers work. Can't seem to get the mono debug extension for VSCode to work though, but it's not connecting at all (can't even pause the program) so I doubt it's related.

**Comment by PJB3005** [comment]
Ok got the debugger to work, it's still not hitting breakpoints.

**Comment by neikeq** [comment]
This was fixed by #18949. For existing projects, you need to manually rename all `<DebugType>full</DebugType>` occurrences in your csproj file to `<DebugType>portable</DebugType>`.
